### PR TITLE
Add check for invalid #[macro_export] arguments

### LINKED ISSUE #107231: #[macro_export(hello, world)]
I tried this code:

```rust
#[macro_export(hello, world)]
macro_rules! foo {
    () => {}
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9c40221c498bdc99beb7022d53fdb086

I expected to see this happen: it shouldn't compile

Instead, this happened: it compiles

### Meta

`rustc --version --verbose`:
```
N/A
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by blyxyas**
Resolves #107231
Sorry if I made something wrong, this is my first contribution to the repo.

### CODE REVIEW

**Comment by compiler-errors** [comment]
two things about the title - 1. it doesn't need the issue number in the title, and 2. can you make it "functional" -- what does it do? e.g. "add a check for invalid macro_export arguments" or something?

File: compiler/rustc_error_messages/locales/en-US/passes.ftl [hunk 1]
```diff
@@ -745,3 +745,7 @@ passes_proc_macro_invalid_abi = proc macro functions may not be `extern "{$abi}"
 passes_proc_macro_unsafe = proc macro functions may not be `unsafe`
 
 passes_skipping_const_checks = skipping const checks
+
+passes_invalid_macro_export_arguments = `{$name}` isn't a valid `#[macro_export]` argument.
+
+passes_invalid_macro_export_arguments_too_many_items = `#[macro_export]` can only take 1 or 0 arguments.
```
Comment (clubby789) [inline compiler/rustc_error_messages/locales/en-US/passes.ftl:751]: Most error messages don't have a period at the end so it's probably best to be consistent here
```suggestion
passes_invalid_macro_export_arguments = `{$name}` isn't a valid `#[macro_export]` argument

passes_invalid_macro_export_arguments_too_many_items = `#[macro_export]` can only take 1 or 0 arguments
```

File: compiler/rustc_passes/src/check_attr.rs [hunk 2]
```diff
@@ -2087,7 +2088,32 @@ impl CheckAttrVisitor<'_> {
...
+        } else if let Some(meta_item_list) = attr.meta_item_list() {
+            if meta_item_list.len() > 1 {
+                self.tcx.emit_spanned_lint(
+                    INVALID_MACRO_EXPORT_ARGUMENTS,
+                    hir_id,
+                    attr.span,
+                    errors::MacroExport::TooManyItems,
+                );
+            } else {
+                if meta_item_list[0].name_or_empty() != sym::local_inner_macros {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/check_attr.rs:2106]: What does this do if you just have `#[macro_export()]`?

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 3]
```diff
@@ -4103,3 +4103,33 @@ declare_lint! {
...
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The only valid list of arguments is `#[macro_export(local_inner_macros)]` or no arguments (`#[macro_export]`), you can't have multiple arguments in a macro export, or invalid ones.
+    ///
+    ///
+    pub INVALID_MACRO_EXPORT_ARGUMENTS,
+    Deny,
+    "There aren't any attributes named \"invalid_parameters\"",
```
Comment (compiler-errors) [inline compiler/rustc_lint_defs/src/builtin.rs:4134]: Why does this message reference `"invalid_parameters"`?

File: compiler/rustc_passes/src/check_attr.rs [hunk 4]
```diff
@@ -2087,7 +2088,32 @@ impl CheckAttrVisitor<'_> {
...
     fn check_macro_export(&self, hir_id: HirId, attr: &Attribute, target: Target) {
         if target != Target::MacroDef {
-            self.tcx.emit_spanned_lint(UNUSED_ATTRIBUTES, hir_id, attr.span, errors::MacroExport);
+            self.tcx.emit_spanned_lint(
+                UNUSED_ATTRIBUTES,
+                hir_id,
+                attr.span,
+                errors::MacroExport::Normal,
+            );
+        } else if let Some(meta_item_list) = attr.meta_item_list() {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/check_attr.rs:2097]: What if it's a `#[macro_export = ".."]`?

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 5]
```diff
@@ -4103,3 +4103,33 @@ declare_lint! {
...
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The only valid list of arguments is `#[macro_export(local_inner_macros)]` or no arguments (`#[macro_export]`), you can't have multiple arguments in a macro export, or invalid ones.
+    ///
+    ///
+    pub INVALID_MACRO_EXPORT_ARGUMENTS,
+    Deny,
```
Comment (compiler-errors) [inline compiler/rustc_lint_defs/src/builtin.rs:4133]: This needs to start out at a warn level, not Deny, or else it'll break a bunch of crates that are using this incorrectly currently.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 6]
```diff
@@ -4103,3 +4103,33 @@ declare_lint! {
     };
     report_in_external_macro
 }
+
+declare_lint! {
+    /// The `invalid_macro_export_arguments` lint detects cases where `#[macro_export]` is being used with the wrong number of arguments, or invalid ones.
```
Comment (compiler-errors) [inline compiler/rustc_lint_defs/src/builtin.rs:4108]:
```suggestion
    /// The `invalid_macro_export_arguments` lint detects cases where `#[macro_export]` is being used with invalid arguments.
```

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 7]
```diff
@@ -4103,3 +4103,33 @@ declare_lint! {
...
+    /// }
+    ///
+    /// #[macro_export(too, many, items)]
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The only valid list of arguments is `#[macro_export(local_inner_macros)]` or no arguments (`#[macro_export]`), you can't have multiple arguments in a macro export, or invalid ones.
```
Comment (compiler-errors) [inline compiler/rustc_lint_defs/src/builtin.rs:4129]:
```suggestion
    /// The only valid list of arguments is `#[macro_export(local_inner_macros)]` or no arguments (`#[macro_export]`).
    /// You can't have multiple arguments in a `#[macro_export(..)]`, or mention arguments other than `local_inner_macros`.
```

Maybe this?

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 5 above)
Comment (blyxyas) [reply compiler/rustc_lint_defs/src/builtin.rs:4133]: Ok
Does this need a future incompatible warn?

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 6 above)
Comment (blyxyas) [reply compiler/rustc_lint_defs/src/builtin.rs:4108]: Fixed in new commit

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 7 above)
Comment (blyxyas) [reply compiler/rustc_lint_defs/src/builtin.rs:4129]: Fix in new commit

**Comment by blyxyas** [comment]
The code is all fixed, but testing gives an error and I don't know why. The stderr doesn't show any diff and cannot see what I'm doing wrong

<details>
<summary>
Output
</summary>

```
failures:

---- [ui] tests/ui/invalid_macro_export_argument.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: "/home/alejandra/git/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/alejandra/git/rust/tests/ui/invalid_macro_export_argument.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/alejandra/git/rust/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/home/alejandra/git/rust/build/x86_64-unknown-linux-gnu/test/ui/invalid_macro_export_argument" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/alejandra/git/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/alejandra/git/rust/build/x86_64-unknown-linux-gnu/test/ui/invalid_macro_export_argument/auxiliary"
stdout: none
--- stderr -------------------------------
warning: `#[macro_export]` can only take 1 or 0 arguments
  --> fake-test-src-base/invalid_macro_export_argument.rs:1:1
   |
LL | #[macro_export(hello, world)] //~ WARN `#[macro_export]` can only take 1 or 0 arguments
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(invalid_macro_export_arguments)]` on by default

warning: `not_local_inner_macros` isn't a valid `#[macro_export]` argument
  --> fake-test-src-base/invalid_macro_export_argument.rs:6:16
   |
LL | #[macro_export(not_local_inner_macros)] //~ WARN `not_local_inner_macros` isn't a valid `#[macro_export]` argument
   |                ^^^^^^^^^^^^^^^^^^^^^^

warning: 2 warnings emitted
------------------------------------------



failures:
    [ui] tests/ui/invalid_macro_export_argument.rs

test result: FAILED. 4 passed; 1 failed; 14429 ignored; 0 measured; 0 filtered out; finished in 0.23s

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Build completed unsuccessfully in 0:00:02
```

<details>

The lint level has been changed to a `Warn` level, so I changed `//~ ERROR` to `//~ WARN` and it still gives me an error (without any diffs).

Is there something that I'm doing wrong?

**Comment by compiler-errors** [comment]
If you make it a warning, then you need to mark the test with `// check-pass`, or else we expect a failure. You can also just add a `#![deny(invalid_macro_export_arguments)]` to make them into an error just for your UI test.

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 5 above)
Comment (compiler-errors) [reply compiler/rustc_lint_defs/src/builtin.rs:4133]: That can be added in a follow-up

File: compiler/rustc_passes/src/check_attr.rs (hunk 4 above)
Comment (blyxyas) [reply compiler/rustc_passes/src/check_attr.rs:2097]: When testing this:

```
error: malformed `macro_export` attribute input
  --> fake-test-src-base/invalid_macro_export_argument.rs:31:1
   |
LL | #[macro_export = "Not a valid value"] //~ WARN `not_local_inner_macros` isn't a valid `#[macro_export]` argument
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
```

It seems like Rust already gives an error in that case.When testing this:

```
error: malformed `macro_export` attribute input
  --> fake-test-src-base/invalid_macro_export_argument.rs:31:1
   |
LL | #[macro_export = "Not a valid value"] //~ WARN `not_local_inner_macros` isn't a valid `#[macro_export]` argument
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
```

It seems like Rust already gives an error in that case.

File: compiler/rustc_lint_defs/src/builtin.rs (hunk 3 above)
Comment (blyxyas) [reply compiler/rustc_lint_defs/src/builtin.rs]: Fixed

**Comment by compiler-errors** [comment]
Can you squash these into one commit? I think this is ready to go after that.

**Comment by compiler-errors** [comment]
Whoops, sorry, forgot to approve.

r? @compiler-errors @bors r+

**Comment by blyxyas** [comment]
@rustbot label: -T-bootstrap
Sorry to Eric Huss for the mention. Im fighting with Git right now trying to resolve the conflict.

**Comment by compiler-errors** [comment]
@blyxyas, I can push to your branch and rebase this if you're having trouble.

**Comment by blyxyas** [comment]
I'd appreciate that a lot. Thanks

**Comment by compiler-errors** [comment]
I'll wait for CI to check this PR, then I can re-approve.

In the future, `git pull --rebase origin master` (or instead of `origin`, whatever your remote for rust-lang/rust.git is called) is the easiest way to rebase past things like this. Submodules can be a pain though.

**Pull request merged**